### PR TITLE
bounty payout v2

### PR DESCRIPTION
**Pull request by johannbarbie**
resolves https://github.com/leapdao/leap-contracts/issues/248

### CODE REVIEW

**Comment by troggy** [comment]
how to use this contract for invoice-like payments?

**Comment by troggy** [comment]
As a hack I would just put in 1 for gardener and everything else into worker share

**Comment by johannbarbie** [comment]
> how to use this contract for invoice-like payments?

a gardener is some-one who sets intention for work to be done. so if some-one initiated and executes work under an invoice, that some-one is the gardener.

no work should be paid out without being checked by at least another pair of eyes, so that some-one approving the invoice is the reviewer.

=> use `payoutNoWorker()` function

**Comment by troggy** [comment]
how to use the repOnly thing?

File: contracts/misc/BountyPayout.sol [hunk 1]
```diff
@@ -33,22 +34,19 @@ contract BountyPayout {
...
 
-  modifier onlyPayer() {
-    require(msg.sender == payerAddr, "Only payer can call");
-    _;
+  function _isRepOnly(uint256 amount) internal returns (bool) {
+    return ((amount & 0x01) == 1);
   }
 
-  function makePayment(address payable _worker, uint256 _amount) internal returns (uint256) {
+  function _makePayment(address payable _worker, uint256 _amount) internal returns (uint256) {
```
Comment (troggy) [inline contracts/misc/BountyPayout.sol]: rename to smth like `makeColonyPayment` ?

File: contracts/misc/BountyPayout.sol [hunk 2]
```diff
@@ -117,64 +123,54 @@ contract BountyPayout {
...
-  ) public onlyPayer {
+  ) public onlyOwner {
     _payout(
-      _gardenerAddr,
-      _gardenerDaiAmount,
-      _workerAddr,
-      _workerDaiAmount,
-      _reviewerAddr,
-      _reviewerDaiAmount,
+      address(bytes20(_gardener)),
```
Comment (troggy) [inline contracts/misc/BountyPayout.sol]: 🤯

**Comment by johannbarbie** [comment]
> `payout` pulls from `msg.sender`, but can be called only by the owner. And there can be just one owner as per `Ownable` implementation. Maybe you wanted to use [Whitelist](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/access/roles/WhitelistedRole.sol) instead?

totally right, i confused it with how `MintableRole` works. `WhitelistedRole` needs an admin though, i chose something in between to fix it.

**Comment by johannbarbie** [comment]
> how to use the repOnly thing? just add 1 DAI ?

yeah, add a DAI-WAI at the end.

**Comment by troggy** [comment]
> WhitelistedRole needs an admin though

admin could be a General Multisig 😎

**Comment by johannbarbie** [comment]
> admin could be a General Multisig

i will put this in before merging.

**Pull request merged**